File: Trapping Rain Water/Trapping Rain Water.java
public class Solution {
    public int trap(int[] A) {
        
        if(A.length <= 2)       return 0;
        
        int[] maxLeft = new int[A.length];
        int[] maxRight = new int[A.length];
        
        maxLeft[0] = 0;
        for(int i = 1; i < A.length; ++i){
            maxLeft[i] = Math.max(maxLeft[i - 1], A[i - 1]);
        }
        
        maxRight[A.length - 1] = 0;
        for(int i = A.length - 2; i >= 0; --i){
            maxRight[i] = Math.max(maxRight[i + 1], A[i + 1]);
        }
        
        int ret = 0;
        for(int i = 0; i < A.length; ++i){
            int min = Math.min(maxLeft[i], maxRight[i]);
            if(min > A[i]){
                ret += min - A[i];
            }
        }
        
        return ret;
    }
}

File: Trapping Rain Water/Trapping Rain Water.cpp
class Solution {
public:
    int trap(int A[], int n) {
        if(!n || (n == 1) || (n == 2))      return 0;
        
        int* maxLeft = new int[n];
        int* maxRight = new int[n];
        
        for(unsigned int i = 0; i < n; ++i){
            maxLeft[i] = 0;
            maxRight[i] = 0;
        }
        
        for(unsigned int i = 1; i < n; ++i){
            maxLeft[i] = (maxLeft[i - 1] > A[i - 1]) ? maxLeft[i - 1] : A[i - 1];
        }
        for(int i = n - 2; i >= 0; --i){
            maxRight[i] = (maxRight[i + 1] > A[i + 1]) ? maxRight[i + 1] : A[i + 1];
        }
        int sum = 0;
        for(unsigned int i = 0; i < n; ++i){
            int diff = ((maxLeft[i] < maxRight[i]) ? maxLeft[i] : maxRight[i]) - A[i];
            if(diff > 0){
                sum += diff;
            }
        }
        return sum;
    }
};
